symfony/console/4.2/bin/console

```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env php
+<?php
+
+use App\Kernel;
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Input\ArgvInput;
+use Symfony\Component\Debug\Debug;
+use Symfony\Component\Dotenv\Dotenv;
+
+set_time_limit(0);
+
+require __DIR__.'/../vendor/autoload.php';
+
+if (!class_exists(Application::class)) {
+    throw new \RuntimeException('You need to add "symfony/framework-bundle" as a Composer dependency.');
+}
+
+$input = new ArgvInput();
+$env = $input->getParameterOption(['--env', '-e'], null, true);
```

Comment:
We need way to know which env we want. I thought about adding a new env var, but as this option already exist I prefer to reuse it.

The fallback is done later.

Reply:
Sure, but I would have kept `$env = $input->getParameterOption(['--env', '-e'], $_SERVER['APP_ENV'] ?? 'dev');` as in https://gist.github.com/ogizanagi/84f6e9c306e1dfac7c66ecffce06b858#file-console-diff-L28

Reply:
It would make the code harder to read because the call `loadForEnv($env)` can change the value `$_SERVER['APP_ENV']`. And if the value change because of the env file, it must be taken into account.

With the current approach, it's explicit that the value can come from the env file.

Reply:
If `$_SERVER['APP_ENV']` is set, then the DotEnv component will never override it, right?

I'm not convinced about the `LOAD_ENV` suggestion, as to me the `APP_ENV` is really meant to replace the `--env` option and nothing else.
The fact the `.env` file also sets the `APP_ENV` is precisely a fallback for when the `APP_ENV` is not explicitly provided. Like any other env var.

Reply:
> If $_SERVER['APP_ENV'] is set, then the DotEnv component will never override it, right?

Well, it's the problem. If `APP_ENV` is defined, DotEnv will never be called. It's intended (in prod real env var should be used). But we assume that if `APP_ENV` is defined, all other mandatory environment variables are defined. So just running `APP_ENV=prod bin/console` will most likely not work, you'll have to run `APP_ENV=prod FOO=BAR FOO2=BAR2 [...] bin/console`. While it's easy in production, and when using Docker or Kubernetes, for most user its not a good a DX.

As we cannot use `APP_ENV` to hint that we need to load a specific env file (because it means, the environnement is already properly configured, you don't need DotEnv), the idea is to introduce `LOAD_ENV` to be this hint.
`LOAD_ENV=prod bin/console` means "load .env.prod" and use it.

`LOAD_ENV` will be a debug tool (that can also be leveraged for people not able to use K8S/Docker etc: just set LOAD_ENV=production in the Apache config, and then you can manage your config using .env.production directly, just like you used to do with parameters.yml).

Another benefit:

`LOAD_ENV=panther bin/phpunit`

```
# .env.panther

DATABASE_URL=mysql://test
```

And you can use a specific DB just for Panther.

wdyt @ogizanagi?

Reply:
> Well, it's the problem. If APP_ENV is defined, DotEnv will never be called. It's intended (in prod real env vars should be used). But we assume that if APP_ENV is defined, all other mandatory environment variables are defined. 

Sure. That's why in [the `DotEnvLoader` I'm using](https://gist.github.com/ogizanagi/84f6e9c306e1dfac7c66ecffce06b858#file-dotenvloader-php-L14-L15), I do only throw when **no `APP_ENV` is provided** and **no DotEnv component is available**, as the DotEnv component won't even be installed on production (`composer install --no-dev`).
Then, even if the `APP_ENV` env var is provided, as soon as the DotEnv component is available, I use it to fulfill other env vars from `.env` files.

So, in production without DotEnv component, `.env` file won't be loaded (_In our case, it these files aren't even part of the production release_).
During development, one can provide the `APP_ENV` env var to choose environment, as well as any other env var if they want, and appropriate `.env.*` files will be loaded, fulfilling but not overriding env vars.

Does it make sense?

I don't have counter-argument about the other benefit you mentioned with `LOAD_ENV`, but I don't know if it'll be that much common.

Reply:
Got it! Let me try.

Reply:
@ogizanagi here we are! Very good idea.
And it even allows to do `APP_ENV=panther bin/phpunit` now.